R/bcf.R

```diff
@@ -281,139 +284,111 @@ bcf <- function(y, z, x_control, x_moderate=x_control, pihat, w = NULL,
+                                 use_mscale = use_muscale, use_bscale = use_tauscale, b_half_normal = TRUE, verbose_sigma=verbose)
+    
+    cat("bcfoverparRcppClean returned to R\n")
+    # Compute Rhat and Rhat_interval for:
+    ## * tau
+    ## * sigma
+    ## * tau for each obs
+    
+    m_post     = muy + sdy*fitbcf$m_post[,order(perm)]
+    tau_post   = sdy*fitbcf$b_post[,order(perm)]
+    
+    list(sigma = sdy*fitbcf$sigma,
+         yhat = muy + sdy*fitbcf$yhat_post[,order(perm)],
+         sdy = sdy,
+         con_sd = con_sd,
+         mod_sd = mod_sd,
+         muy = muy,
+         mu  = m_post,
+         tau = tau_post,
+         mu_scale = fitbcf$msd*sdy,
+         tau_scale = fitbcf$bsd*sdy,
+         perm = perm,
+         include_pi = include_pi,
+         random_seed=this_seed
+    )
+    
+  }
 
-  fits <- foreach::foreach(i = 1:nchains) %dopar% {
-    library(bcf2)
-
-    bcffit <- bcf(y, z, x_control, x_moderate=x_control, pihat, w, 
-                n_threads,
-                random_seed,
-                nburn, nsim, nthin, update_interval,
-                ntree_control,
-                sd_control,
-                base_control,
-                power_control,
-                ntree_moderate,
-                sd_moderate,
-                base_moderate,
-                power_moderate,
-                nu, lambda, sigq, sighat,
-                include_pi, use_muscale, use_tauscale, verbose
-) 
+  sigma = c()
+  mu_scale = c()
+  tau_scale = c()
+  
+  chain = c()
+  iteration = c()
+  
+  yhat = c()
+  mu   = c()
+  tau  = c()
+  
+  
+  n_iter = length(chain_out[[1]]$sigma)
+  
+  for (iChain in 1:n_chains){
+    sigma       = c(sigma,     chain_out[[iChain]]$sigma)
+    mu_scale    = c(mu_scale,  chain_out[[iChain]]$mu_scale)
+    tau_scale   = c(tau_scale, chain_out[[iChain]]$tau_scale)
+    
+    
+    chain       = c(chain,     rep(iChain,n_iter))
+    iteration   = c(iteration, 1:n_iter)
+    
+    yhat = rbind(yhat, chain_out[[iChain]]$yhat)
+    mu   = rbind(mu,   chain_out[[iChain]]$mu)
+    tau  = rbind(tau,  chain_out[[iChain]]$tau)
+    
+    if(chain_out[[iChain]]$sdy        != chain_out[[1]]$sdy)        stop("sdy not consistent between chains for no reason")
+    if(chain_out[[iChain]]$con_sd     != chain_out[[1]]$con_sd)     stop("con_sd not consistent between chains for no reason")
+    if(chain_out[[iChain]]$mod_sd     != chain_out[[1]]$mod_sd)     stop("mod_sd not consistent between chains for no reason")
+    if(chain_out[[iChain]]$muy        != chain_out[[1]]$muy)        stop("muy not consistent between chains for no reason")
+    if(chain_out[[iChain]]$include_pi != chain_out[[1]]$include_pi) stop("include_pi not consistent between chains for no reason")
+    if(any(chain_out[[iChain]]$perm   != chain_out[[1]]$perm))      stop("perm not consistent between chains for no reason")
   }
```

Comment:
@cdelannoy  Is there a better "R" way to do this thing where I concatenate the and rbind certain stuff? It seems like a lot of names hardcoded